Feature/get opportunities

rohithmsr opened the pull request:
Fixes #16 
### Description

1. Implemented GET Opportunities, if no query parameter for 'type' is given, it returns all the opportunities
2. changed bodyparser.json() to express.json() middleware as an alternative for bodyparser.json since the latter is deprecated 
3. Updated the documentation

rohithmsr committed:
Add GET Oppotunities With Filter Based on Opportunity Type

rohithmsr committed:
Update API Documentation for GET Opportunities

abdus commented on controllers/opportunity/index.js:
```diff
@@ -41,6 +41,25 @@ class opportunityController {
+      return res.status(200).json({
+        data: fetchedOpportunities,
+      });
+    } catch (err) {
+      console.log('ERROR IN getOpportunities OpportunityController', err);
+      return res.status(400).send(err);
+    }
+  }
```
do you think sending raw error to front-end would be a good idea?

rohithmsr replied:
As of now, we shall send responses in a format like the below,

**For successful request**
```
{
   "status": "success",
   "data": {
          [opportuities]
    }
}
```

**For unsuccessful request**
```
{
   "status": "fail",
   "error": {
          [error that is thrown]
    }
}
```

If you have any suggestions regarding this, I'll change. I have no idea of what to send for responses for a standard API. @abdus

Manvityagi replied:
CC: @abdus

rohithmsr replied:
Any Idea on this ? @abdus @Manvityagi 
Or should I go with the above mentioned format?

Manvityagi replied:
Let me check what standard will we follow and get back to you.

abdus replied:
@rohithmsr this looks good. you can go ahead with this format. just make sure that you process the error message before sending it to the client. for example, a `duplicate field` error thrown by mongoose would not make much sense to the front-end people. Do you understand what I am trying to say?

rohithmsr replied:
Hey @abdus , as of now,
 if have just added status and data fields,
 if error, there will be status and error fields as mentioned above,  in which, inside error key, raw error will be its value. Once we implement some more features, i guess it will be easy to identify errors like ValidationError, duplicate error etc. 

Please!! not as of now, I think you understood what I said @abdus

rohithmsr replied:
If I try to make error in spelling of query param , for example, ?blablabla=blabla, still it returns 0 results, I can write a code to check for the query parameters quickly, but in the future if I want to implement pagination, aliasing etc, the code may break.

Even if I try making errors in URL, it says '/Cannot GET resource/' . This error handling is out of scope for this issue, it needs 
an separate issue

If I try to make error in query param like ?type=blablabla, it returns 0 results which is not an error and it is a successful request.  

Please!! I hope you got it @abdus @Manvityagi 
My suggestion is, we shall take care of error handling in a controller folder after some more features are added.

abdus commented on routes/opportunity.js:
```diff
@@ -24,11 +24,16 @@ const opportunityManager = new OpportunityManager(),
-  console.log('Inside Route');
-  opportunityController.createOpportunity(req, res);
-});
+router
+  .get('/', (req, res) => {
+    console.log('Inside GET Route');
+    opportunityController.getOpportunities(req, res);
+  })
+  .post('/', (req, res) => {
+    console.log('Inside Route');
+    opportunityController.createOpportunity(req, res);
+  });
```
please document the API endpoint using Swagger. Refer to Readme for a tutorial.

rohithmsr replied:
It comes under separate issue right? Sorry to ask this as a budding open source contributor, if you have this as a separate issue I can get points for GSSOC. @abdus

Manvityagi replied:
Usually, API Documentation comes in handy with the same issue, but let me discuss with @abdus and get back to you on this.

rohithmsr replied:
There is a separate issue #19  for Documentation for POST API on swagger.
If it was LEVEL0 issue, I would have not even asked you and done this along with #16 issue. But the issue #19 is of LEVEL2 and doing this as a separate issue would give me some more points. That's why asked @Manvityagi

Manvityagi replied:
@rohithmsr The previous API was added as a part of the basic setup, so swagger documentation wasn't added to it. 
As per the standards in the ideal case, the person who writes the API shall also write the swagger doc for the same API. 

I am increasing the LEVEL of this API from 2 to 3, so please also commit the documentation of this endpoint using Swagger. 

Swagger is easy, You will understand it at once, if you face any problems, feel free to ping on discord channel #opportunity-calendar-backend. 

CC: @abdus

rohithmsr replied:
Sure , will work on it

rohithmsr commented:
There is an open issue #19 for POST . So when shall I start my documentation part for GET on Swagger? Is the work #19  in progress?

Manvityagi commented:
> There is an open issue #19 for POST . So when shall I start my documentation part for GET on Swagger? Is the work #19 in progress?

Your work is independent of that issue. 
Swagger setup has been done already by our end. 
You write the yaml file. 
If there will be merge-conflicts, the person making issue #19 will handle it.

rohithmsr committed:
Add Swagger Docs for GET Opportunities

rohithmsr commented:
Hey, I have added the Swagger Documentation for GET Request. And I have also added formatting of response I mentioned above!!

I hope this pull request will be merged 🤞🤞

@abdus @Manvityagi

Manvityagi commented:
Merged 🎉🎉🥳🥳
